R/table_parts.R

```diff
@@ -253,32 +256,75 @@ tab_row_group <- function(data,
 #' @export
 tab_spanner <- function(data,
                         label,
-                        columns) {
+                        columns,
+                        gather = TRUE) {
   checkmate::assert_character(
     label, len = 1, any.missing = FALSE, null.ok = FALSE)
 
-  # If using the `vars()` helper, get the columns as a character vector
-  if (inherits(columns, "quosures")) {
-    columns <- columns %>% lapply(`[[`, 2) %>% as.character()
-  }
+  data_df <- as.data.frame(data)
+  colnames <- colnames(data_df)
+
+  columns <- enquo(columns)
+
+  resolved_columns <-
+    resolve_vars(var_expr = columns, var_names = colnames, data_df = data_df)
+
+  # Translate the column indices to column names
+  resolved_columns <- colnames[resolved_columns]
 
   # Filter the vector of column names by the
   # column names actually in `input_df`
-  columns <- columns[which(columns %in% colnames(data))]
+  resolved_columns <- resolved_columns[which(resolved_columns %in% colnames)]
 
-  if (length(columns) == 0) {
+  if (length(resolved_columns) == 0) {
     return(data)
   }
 
   grp_labels <- attr(data, "grp_labels", exact = TRUE)
 
-  for (i in seq(columns)) {
-    grp_labels[[columns[i]]] <- label
+  for (i in seq(resolved_columns)) {
+    grp_labels[[resolved_columns[i]]] <- label
   }
 
   # Set the `grp_labels` attr with the `grp_labels` object
   attr(data, "grp_labels") <- grp_labels
 
+  # Gather columns not part of the group of columns under
+  # the spanner heading
+  if (gather && length(resolved_columns) > 1) {
+
+    # Extract the internal `boxh_df` table
+    boxh_df <- attr(data, "boxh_df", exact = TRUE)
+
+    # Get the sequence of columns available in `boxh_df`
+    all_columns <- colnames(boxh_df)
+
+    # Get the vector positions of the `columns` in
+    # `all_columns`
+    matching_vec <-
+      match(resolved_columns, all_columns) %>%
+      sort() %>%
+      unique()
+
+    # Issue a warning if duplicate values were given
+    # in resolved columns
+    if (any(duplicated(resolved_columns))) {
```

Comment:
Seems like this check should be outside the conditional?

Reply:
Yes. But, also, duplicates won't happen anymore because of the use of `resolve_vars()`. So I just took it out (after sanity-checking this thoroughly).